File: Programmers/cpp/LEVEL_2_무인도_여행.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int dx[] = { 0, 0, 1, -1 };
int dy[] = { 1, -1, 0, 0 };

int bfs(vector<vector<int>>& map, vector<vector<bool>>& visit, int a, int b) {
	queue<pair<int, int>> q;
	q.push({ a, b });
	visit[a][b] = true;

	int sum = 0;

	while (!q.empty()) {
		int x = q.front().first;
		int y = q.front().second;
		q.pop();

		sum += map[x][y];

		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];

			if (nx < 0 || ny < 0 || nx >= map.size() || ny >= map[0].size()) {
				continue;
			}

			if (map[nx][ny] != -1 && visit[nx][ny] == false) {
				q.push({ nx, ny });
				visit[nx][ny] = true;
			}
		}
	}

	return sum;
}

vector<int> solution(vector<string> maps) {
	vector<int> answer;

	vector<vector<int>> map(maps.size(), vector<int>(maps[0].length()));
	vector<vector<bool>> visit(maps.size(), vector<bool>(maps[0].length(), false));

	for (int i = 0; i < maps.size(); i++) {
		for (int j = 0; j < maps[i].length(); j++) {
			if (maps[i][j] == 'X') {
				map[i][j] = -1;
				continue;
			}

			map[i][j] = maps[i][j] - '0';
		}
	}

	for (int i = 0; i < map.size(); i++) {
		for (int j = 0; j < map[i].size(); j++) {
			if (map[i][j] != -1 && visit[i][j] == false) {
				int n = bfs(map, visit, i, j);
				answer.push_back(n);
			}
		}
	}

	if (answer.empty()) {
		answer.push_back(-1);
		return answer;
	}

	sort(answer.begin(), answer.end());

	return answer;
}

File: Programmers/java/LEVEL_2_무인도_여행.java
import java.util.*;

class Solution {
    int[] dx = {0, 0, -1, 1};
    int[] dy = {-1, 1, 0, 0};

    private int bfs(int[][] map, boolean[][] visit, int a, int b) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{a, b});
        visit[a][b] = true;

        int sum = 0;

        while (!q.isEmpty()) {
            int[] qx = q.poll();
            int x = qx[0];
            int y = qx[1];

            sum += map[x][y];

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || ny < 0 || nx >= map.length || ny >= map[0].length) {
                    continue;
                }

                if (map[nx][ny] != -1 && visit[nx][ny] == false) {
                    q.add(new int[]{nx, ny});
                    visit[nx][ny] = true;
                }
            }
        }

        return sum;
    }

    public int[] solution(String[] maps) {
        int[] answer = {};
        List<Integer> lands = new ArrayList<>();

        int[][] map = new int[maps.length][maps[0].length()];
        boolean[][] visit = new boolean[maps.length][maps[0].length()];

        for (int i = 0; i < maps.length; i++) {
            for (int j = 0; j < maps[i].length(); j++) {
                if (maps[i].charAt(j) == 'X') {
                    map[i][j] = -1;
                    continue;
                }

                map[i][j] = maps[i].charAt(j) - '0';
            }
        }

        for (int i = 0; i < map.length; i++) {
            for (int j = 0; j < map[i].length; j++) {
                if (map[i][j] != -1 && visit[i][j] == false) {
                    int n = bfs(map, visit, i, j);
                    lands.add(n);
                }
            }
        }

        if (lands.isEmpty()) {
            answer = new int[]{-1};
            return answer;
        }

        answer = new int[lands.size()];

        for (int i = 0; i < lands.size(); i++) {
            answer[i] = lands.get(i);
        }

        Arrays.sort(answer);

        return answer;
    }
}